src/com/intellij/idea/plugin/hybris/properties/impl/PropertiesServiceImpl.kt

```diff
@@ -142,52 +104,114 @@ class PropertiesServiceImpl(val project: Project) : PropertiesService {
+                nestedValue?.let {
+                    var newValue = nestedValue
+                    if (nestedValue.contains(nestedPropertyPrefix)) {
+                        replacePlaceholder(result, nestedKey, nestedValue)
+                        newValue = result[nestedKey]
+                    }
+
+                    result[key] = value.replace(placeholder, newValue ?: "")
+                }
+            }
         }
-        val matchedFiles = dir.listFiles { _, name -> optionalPropertiesFilePattern.matcher(name).matches() }
-            ?: return
-        val propertyFiles = TreeMap<String, File>()
-        Arrays.stream(matchedFiles).forEach { file -> propertyFiles[file.name] = file }
-
-        propertyFiles.values.forEach { file ->
-            val virtualFile = LocalFileSystem.getInstance().findFileByIoFile(file)
-            if (virtualFile == null || !virtualFile.exists()) {
+    }
+
+    private fun loadHybrisOptionalConfigDir(result: LinkedHashMap<String, IProperty>) {
+        var optDir = result[HybrisConstants.PROPERTY_OPTIONAL_CONFIG_DIR]?.value
+
+        val envOptConfigDir = System.getenv("HYBRIS_OPT_CONFIG_DIR")
+        if (envOptConfigDir != null) {
+            optDir = envOptConfigDir;
+        }
+
+        optDir?.let {
+            val dir = File(it)
+            if (!dir.isDirectory) {
                 return
             }
-            val psiFile = PsiManager.getInstance(project).findFile(virtualFile)
-            if (psiFile is PropertiesFile) {
-                addPropertyFile(result, psiFile as PropertiesFile?)
+
+            val matchedFiles = dir.listFiles { _, name -> optionalPropertiesFilePattern.matcher(name).matches() }
+                ?: return
+            val propertyFiles = TreeMap<String, File>()
+            Arrays.stream(matchedFiles).forEach { file -> propertyFiles[file.name] = file }
+
+            propertyFiles.values.forEach { file ->
+                addPropertyFile(result, toPropertiesFile(file))
+            }
+        }
+    }
+
+    private fun loadHybrisRuntimeProperties(result: LinkedHashMap<String, IProperty>) {
+        val envVar: String = System.getenv("HYBRIS_RUNTIME_PROPERTIES") ?: ""
+        if (!envVar.isBlank()) {
+            var propertiesFile = File(envVar);
+            propertiesFile?.let { file: File ->
+                addPropertyFile(result, toPropertiesFile(file))
             }
+
         }
     }
 
+    private fun toPropertiesFile(file: File): PropertiesFile? {
+        val virtualFile: VirtualFile? = LocalFileSystem.getInstance().findFileByIoFile(file)
+        if (virtualFile == null || !virtualFile.exists()) {
+            return null
+        }
+        val psiFile: PsiFile? = PsiManager.getInstance(project).findFile(virtualFile)
+        if (psiFile is PropertiesFile) {
+            return psiFile
+        }
+        return null
+    }
+
+
     private fun addPropertyFile(result: LinkedHashMap<String, IProperty>, propertiesFile: PropertiesFile?) {
```

Comment:
with all refactoring efforts it would be nice to make `propertiesFile` non nullable